### PR TITLE
Allow _FillValue and missing_value to differ (Fixes #1749)

### LINKED ISSUE #1749: _FillValue and missing_value not allowed for the same variable
#### Problem description

Currently opening the file which has a variable with both _FillValue and missing_value defined, results in a ValueError "'Discovered conflicting _FillValue and missing_value...", which can be overridden by setting `decode_cf=False` 

Code in `xarray/conventions.py` have comments that `missing_value is deprecated`, however this seems not to be true anymore, see [ticket #67](https://cf-trac.llnl.gov/trac/ticket/67) in http://cfconventions.org/cf-conventions/cf-conventions.html

Also, see description of meanings of _FillValue and missing_value here: 
https://github.com/Unidata/netcdf4-python/issues/94
"Note that missing_value and _FillValue are very different. _FillValue is used by the C library to fill in values in variables that have not yet been written to. missing_value is meant to represent values set by the user to represent data points that do not exist."

#### Expected Output
Accept both _FillValue and missing_value and if both present, mask output only for _FillValue

#### Output of ``xr.show_versions()``
0.9.6

### PR DESCRIPTION
**Pull request by dopplershift**
The CF standard permits both values, and them to have different values, so we should not be treating this as an error--just mask out all of them.

 - [x] Closes #1749 (remove if there is no corresponding issue, which should only be the case for minor changes)
 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

### CODE REVIEW

**Comment by dopplershift** [comment]
Test failures look unrelated--they also happen to me locally. Installing SciPy 1.0.0 locally fixes them, so I'm guessing they're caused by SciPy 1.0.1.

**Comment by shoyer** [comment]
please merge in master, which fixes the failing tests.

**Comment by dopplershift** [comment]
Rebased on current master.

**Comment by dopplershift** [comment]
Fixed flake8 (unnecessary import)

**Comment by dopplershift** [comment]
Reworked due to corner case with PyNio on Python 2.7. Final solution looks simpler to me.

File: xarray/coding/variables.py [hunk 1]
```diff
@@ -152,26 +152,12 @@ def encode(self, variable, name=None):
...
-                fv for fv in np.ravel(raw_fill_value) if not pd.isnull(fv)]
+        raw_fill_values = [pop_to(attrs, encoding, attr, name=name)
+                           for attr in ('missing_value', '_FillValue')]
+        if raw_fill_values:
+            encoded_fill_values = [fv for option in raw_fill_values
+                                   for fv in np.ravel(option)
+                                   if not pd.isnull(fv)]
 
             if len(encoded_fill_values) > 1:
                 warnings.warn("variable {!r} has multiple fill values {}, "
```
Comment (shoyer) [inline xarray/coding/variables.py]: Is this warning this useful? Or should it be removed?

File: xarray/tests/test_conventions.py [hunk 2]
```diff
@@ -156,12 +156,16 @@ def test(self):
 
 
 def test_decode_cf_with_conflicting_fill_missing_value():
-    var = Variable(['t'], np.arange(10),
+    expected = Variable(['t'], [np.nan, np.nan, 2], {'units': 'foobar'})
+    expected[[0, 1]] = np.nan
```
Comment (shoyer) [inline xarray/tests/test_conventions.py]: I think this line is unnecessary, since you already set these values to NaN when you construct the variable.

Comment (dopplershift) [reply xarray/tests/test_conventions.py]: Oops, leftover from trying to get the thing to work. 😁

File: xarray/coding/variables.py (hunk 1 above)
Comment (dopplershift) [reply xarray/coding/variables.py]: Personally, I don't need it. I'm not sure if others would like to be made of this case, though. I think my thoughts on leaving it were that it might be good to know since it would affect round-tripping.

One thing I notice nowis that this will now fire if `missing_value` and `_FillValue` are both present, regardless of their values. That's probably a bad thing. So I guess I'm 👍 on removing it.

Comment (shoyer) [reply xarray/coding/variables.py]: Yeah, I think it probably makes sense to keep if it's an indication that round-tripping will likely fail.

Comment (dopplershift) [reply xarray/coding/variables.py]: Ok, leaving the warning. I updated the comprehensions to make a set, so that should suppress false positives with the warning.

**Comment by shoyer** [comment]
thanks!

**Comment by himikof** [comment]
It seems that this PR exposed a new encoding key 'missing_value', and this makes the encoding fail to round-trip, as it is not explicitly allowed in `xarray.backends.netCDF4_._extract_nc4_variable_encoding` function (similar to #1869).
This leads to the following error when saving with an explicitly preserved encoding: 
ValueError: unexpected encoding parameters for 'netCDF4' backend:  ['missing_value']

**Comment by dopplershift** [comment]
@shoyer Happy to put in a PR to address...provided you can tell me what that should actually look like. 😁

**Comment by shoyer** [comment]
Well, we're not going to perfectly round-trip data that uses both `_FillValue` *and* `missing_value`. That isn't possible with xarray's data model.

But we should certainly not raise errors when round-tripping. I see a few reasonable options:
- Don't store `missing_value` in encoding. This would entail removing the use of `pop_to` with `missing_value` in `decode()`.
- Add some logic to `encode()` to check which of `_FillValue` and/or `missing_value` are defined. Use the attribute is defined and write it to attribute. If both are defined, ignore/drop `missing_value`.

As a regression test, we should verify that `code.encode(coder.decode(var)) == var` (including explicitly checking `encoding`).
